File: uva/borrowers.py
import re, sys, collections

p_book = re.compile("\"([^\"]+)\" by (.+)")
p_cmd = re.compile("(\w+)( \"([^\"]+)\")?")

class Book():
    def __init__(self, title, author):
        self.key = (author, title)
        self.title = title
        self.author = author
        self.avail = True

if __name__ == "__main__":
    books = []
    while True:
        line = input()
        if line == "END":
            break
        books.append(Book(*p_book.match(line).group(1,2)))
    books.sort(key=lambda b: b.key)
    books_dict = collections.OrderedDict(zip((b.title for b in books), range(len(books))))

    lines = []
    books_return = []

    while True:
        cmd, title = p_cmd.match(input()).group(1,3)
        if cmd == "BORROW":
            i = books_dict[title]
            if books[i].avail:
                books[i].avail = False
            else:
                books_return.remove(i)
        elif cmd == "RETURN":
            books_return.append(books_dict[title])
        elif cmd == "SHELVE":
            books_return.sort()
            for i in books_return:
                books[i].avail = True
                first = True
                if i > 0:
                    for book in books[i-1::-1]:
                        if book.avail:
                            lines.append("Put \"{}\" after \"{}\"".format(books[i].title, book.title))
                            first = False
                            break
                if first:
                    lines.append("Put \"{}\" first".format(books[i].title))
            lines.append("END")
            books_return = []
        elif cmd == "END":
            break

    print('\n'.join(lines))
    path = sys.path[0] + "\\borrowers.out"
    with open(path, "w") as file:
        file.write("\n".join(lines))

File: uva/borrowers.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
  set<pair<string, string> > books;
  set<pair<string, string> > books_return;
  map<string, string> authors;
  string line;
  smatch sm;
  regex e1("\"([^\"]+)\" by (.+)");
  while (true){
    getline(cin, line);
    if (line == "END") break;
    regex_match(line, sm, e1);
    books.emplace(sm[2], sm[1]);
    authors[sm[1]] = sm[2];
  }

  regex e2("(\\w+)(?: \"([^\"]+)\")?");
  while (true){
    getline(cin, line);
    if (line == "END") break;
    regex_match(line, sm, e2);
    if (sm[1] == "BORROW"){
      pair<string, string> b(authors[sm[2]], sm[2]);
      books_return.erase(b) || books.erase(b);
    } else if (sm[1] == "RETURN"){
      books_return.emplace(authors[sm[2]], sm[2]);
    } else if (sm[1] == "SHELVE"){
      for (auto it=books_return.begin(); it!=books_return.end(); ++it){
        auto prev = books.emplace(it->first, it->second);
        if (prev.first != books.begin()){
          --prev.first;
          printf("Put \"%s\" after \"%s\"\n", it->second.c_str(), prev.first->second.c_str());
        } else {
          printf("Put \"%s\" first\n", it->second.c_str());
        }
      }
      cout << "END\n";
      books_return.clear();
    }
  }

  return 0;
}
